Remove factor of half in data misfits and regularizations

lheagy opened the pull request:
#### Summary
Remove the factor of 1/2 in front of data misfit and regularization terms. This has caused many headaches as outlined in #763. 

One important note:
- the Joint Total Variation never had a 1/2 in front. The documentation and the implementation did not match each other (the documentation stated there was a factor of 1/2, but in the code, there wasn't). So this will change _will_ require that scaling factors used in front of the joint total variation regularization be multiplied by a factor of 2 to obtain the same result if a fixed value was used (cc @jcapriot: I would appreciate a double-check from you on this if you don't mind!)  

#### Todos and input! 
- [x] I would appreciate input from @santisoler and / or @thast on the PGI updates 
- [x] @fourndo : the sparse norms were more straightforward, but I would still appreciate your eyes on this
- [x] if anyone is willing to run some inversion examples before / after this change, that would be a much appreciated test! We should still start at approximately the same beta when using the beta estimator. I will test out a few examples, but it would not hurt to have others test this out too! 

#### PR Checklist
* [x] If this is a work in progress PR, set as a Draft PR
* [x] Linted my code according to the [style guides](https://docs.simpeg.xyz/content/getting_started/contributing/code-style.html).
* [x] Added [tests](https://docs.simpeg.xyz/content/getting_started/practices.html#testing) to verify changes to the code.
* [x] Added necessary documentation to any new functions/classes following the
      expect [style](https://docs.simpeg.xyz/content/getting_started/practices.html#documentation).
* [x] Marked as ready for review (if this is was a draft PR), and converted 
      to a Pull Request
* [x] Tagged ``@simpeg/simpeg-developers`` when ready for review.

#### Reference issue
This will close #763 

#### What does this implement/fix?
- remove the factor of 1/2 in front of data misfit, regularization terms
- update the target misfit directive to also remove the factor of 1/2
- update documentation to remove factor of 1/2 
- update derivatives to now have a factor of 2
- update tests accordingly 

#### Summary for Squash and Merge

Remove the factor of 1/2 in front of data misfit and regularization terms. Add a factor of 2 to their derivatives. Update target misfit directives to account for the missing 1/2 factor. Update documentation to remove the factor of 1/2. Update tests accordingly (#1332 for PGI).

domfournier reviewed (approved):
Looks straight "forward" to me ;). 
Glad to see this coming in.

santisoler reviewed (changes requested):
Thanks @lheagy for taking care of this. I think it's looking great and I don't have any doubts that we'll have the 1/2 factor removed for the next release.

I just pushed a few changes. Most of them remove the 1/2 factor in docstrings of regularizations, data misfit and objective function classes. I also removed two `0.5` in the `PairedBetaSchedule` directive. But I would love if anyone can take a second look at those.

I removed the 1/2 in the docstrings of `JointTotalVariation`, but I'm not sure if the implementation should also be changed. 👀 needed there!

I left a few comments in your code, and here I linked to some non-edited bits that might need some work too.

Let me know what do you think! I can keep helping to get this merged.

---

Should we add a factor of 2 in this intermediate return?
https://github.com/simpeg/simpeg/blob/b5dc70851e045275644fb0e6c45f9e8d82080382/SimPEG/regularization/pgi.py#L844-L859

---

I think we also need to remove these `0.5`, but I'm not 100% sure. There might be some other bits in those tests that we should probably change too:
https://github.com/simpeg/simpeg/blob/eef5e70557d2f9407485bef0542863c9d13ef43f/tests/em/em1d/test_EM1D_FD_jac_layers.py#L118
https://github.com/simpeg/simpeg/blob/eef5e70557d2f9407485bef0542863c9d13ef43f/tests/em/em1d/test_EM1D_FD_jac_layers.py#L158
https://github.com/simpeg/simpeg/blob/eef5e70557d2f9407485bef0542863c9d13ef43f/tests/em/em1d/test_EM1D_FD_jac_layers.py#L277
https://github.com/simpeg/simpeg/blob/eef5e70557d2f9407485bef0542863c9d13ef43f/tests/em/em1d/test_EM1D_FD_jac_layers.py#L317
https://github.com/simpeg/simpeg/blob/eef5e70557d2f9407485bef0542863c9d13ef43f/tests/em/em1d/test_EM1D_FD_jac_layers.py#L453

---

We need to update this example too, right?
https://github.com/simpeg/simpeg/blob/eef5e70557d2f9407485bef0542863c9d13ef43f/examples/20-published/plot_tomo_joint_with_volume.py#L39-L83

jcapriot commented:
The `CrossGradient` and the `JointTotalVariation` both look good to me. I must've forgotten to put it in there at the start, as the derivative didn't need a 2x multiplier (the sqrt(x**2) term essentially). The docstring also looks improved, it would appear that it got erroneously added in by dccowan during #1225 and I missed it on review.

jcapriot commented:
> The `CrossGradient` and the `JointTotalVariation` both look good to me. I must've forgotten to put it in there at the start, as the derivative didn't need a 2x multiplier (the sqrt(x**2) term essentially). The docstring also looks improved, it would appear that it got erroneously added in by dccowan during #1225 and I missed it on review.

Ah, I found the "factor of 1/2" in there, oddly enough. There is an averaging operation that is done inside the square root for JTV that averages the squared values of the gradient vectors instead of just summing them, before computing their magnitudes. I think it actually works out to being a total factor of (1/4) due to the averaging operation being inside the outer most square root though. I don't know if it's technically there though. (If it had been a squared L2 norm of the vector gradients' amplitudes it would've been a 1/2).

santisoler commented:
I just ran an inversion on synthetic gravity data with SimPEG `v0.20.0` and with the latest changes in this branch (`54da8d025b00bba9b7641f1349f5cee1397fc891`) and got comparable results. They don't pass a `np.allclose()`

Here are the printouts of the inversion:

With `v0.20.0`:

<details>

```
SimPEG.InvProblem will set Regularization.reference_model to m0.
SimPEG.InvProblem will set Regularization.reference_model to m0.
SimPEG.InvProblem will set Regularization.reference_model to m0.
SimPEG.InvProblem will set Regularization.reference_model to m0.
SimPEG.InvProblem will set Regularization.reference_model to m0.
SimPEG.InvProblem will set Regularization.reference_model to m0.
SimPEG.InvProblem will set Regularization.reference_model to m0.

                    SimPEG.InvProblem is setting bfgsH0 to the inverse of the eval2Deriv.
                    ***Done using the default solver Pardiso and no solver_opts.***
                    
model has any nan: 0
=============================== Projected GNCG ===============================
  #     beta     phi_d     phi_m       f      |proj(x-g)-x|  LS    Comment   
-----------------------------------------------------------------------------
x0 has any nan: 0
   0  1.53e-04  2.44e+03  0.00e+00  2.44e+03    1.42e+02      0              
   1  5.09e-05  1.15e+03  2.91e+06  1.30e+03    1.23e+02      0              
   2  1.70e-05  7.09e+02  7.90e+06  8.43e+02    9.78e+01      0   Skip BFGS  
------------------------- STOP! -------------------------
1 : |fc-fOld| = 0.0000e+00 <= tolF*(1+|f0|) = 2.4416e+02
0 : |xc-x_last| = 4.5715e-01 <= tolX*(1+|x0|) = 1.0000e-01
0 : |proj(x-g)-x|    = 9.7600e+01 <= tolG          = 1.0000e-01
0 : |proj(x-g)-x|    = 9.7600e+01 <= 1e3*eps       = 1.0000e-02
0 : maxIter   =      20    <= iter          =      3
------------------------- DONE! -------------------------
CPU times: user 30.8 s, sys: 17.5 s, total: 48.2 s
Wall time: 11.1 s
```

</details>


With `54da8d025b00bba9b7641f1349f5cee1397fc891`:

<details>

```
SimPEG.InvProblem will set Regularization.reference_model to m0.
SimPEG.InvProblem will set Regularization.reference_model to m0.
SimPEG.InvProblem will set Regularization.reference_model to m0.
SimPEG.InvProblem will set Regularization.reference_model to m0.
SimPEG.InvProblem will set Regularization.reference_model to m0.
SimPEG.InvProblem will set Regularization.reference_model to m0.
SimPEG.InvProblem will set Regularization.reference_model to m0.

                    SimPEG.InvProblem is setting bfgsH0 to the inverse of the eval2Deriv.
                    ***Done using the default solver Pardiso and no solver_opts.***
                    
model has any nan: 0
=============================== Projected GNCG ===============================
  #     beta     phi_d     phi_m       f      |proj(x-g)-x|  LS    Comment   
-----------------------------------------------------------------------------
x0 has any nan: 0
   0  1.50e-04  4.88e+03  0.00e+00  4.88e+03    1.44e+02      0              
   1  5.01e-05  2.28e+03  5.93e+06  2.58e+03    1.36e+02      0              
   2  1.67e-05  1.41e+03  1.60e+07  1.68e+03    1.18e+02      0   Skip BFGS  
------------------------- STOP! -------------------------
1 : |fc-fOld| = 0.0000e+00 <= tolF*(1+|f0|) = 4.8822e+02
0 : |xc-x_last| = 4.5579e-01 <= tolX*(1+|x0|) = 1.0000e-01
0 : |proj(x-g)-x|    = 1.1765e+02 <= tolG          = 1.0000e-01
0 : |proj(x-g)-x|    = 1.1765e+02 <= 1e3*eps       = 1.0000e-02
0 : maxIter   =      20    <= iter          =      3
------------------------- DONE! -------------------------
CPU times: user 32.5 s, sys: 19.4 s, total: 51.9 s
Wall time: 11.6 s
```

</details>

santisoler commented:
I just ran an inversion on synthetic gravity data with SimPEG `v0.20.0` and with the latest changes in this branch (`54da8d025b00bba9b7641f1349f5cee1397fc891`) and got comparable results. They pass a `np.allclose()` after setting an `atol` to avoid comparing values close to zero.

Here are the printouts of the inversion:

With `v0.20.0`:

<details>

```
SimPEG.InvProblem will set Regularization.reference_model to m0.
SimPEG.InvProblem will set Regularization.reference_model to m0.
SimPEG.InvProblem will set Regularization.reference_model to m0.
SimPEG.InvProblem will set Regularization.reference_model to m0.
SimPEG.InvProblem will set Regularization.reference_model to m0.
SimPEG.InvProblem will set Regularization.reference_model to m0.
SimPEG.InvProblem will set Regularization.reference_model to m0.

                    SimPEG.InvProblem is setting bfgsH0 to the inverse of the eval2Deriv.
                    ***Done using the default solver Pardiso and no solver_opts.***
                    
model has any nan: 0
=============================== Projected GNCG ===============================
  #     beta     phi_d     phi_m       f      |proj(x-g)-x|  LS    Comment   
-----------------------------------------------------------------------------
x0 has any nan: 0
   0  1.53e-04  2.44e+03  0.00e+00  2.44e+03    1.42e+02      0              
   1  5.09e-05  1.15e+03  2.91e+06  1.30e+03    1.23e+02      0              
   2  1.70e-05  7.09e+02  7.90e+06  8.43e+02    9.78e+01      0   Skip BFGS  
------------------------- STOP! -------------------------
1 : |fc-fOld| = 0.0000e+00 <= tolF*(1+|f0|) = 2.4416e+02
0 : |xc-x_last| = 4.5715e-01 <= tolX*(1+|x0|) = 1.0000e-01
0 : |proj(x-g)-x|    = 9.7600e+01 <= tolG          = 1.0000e-01
0 : |proj(x-g)-x|    = 9.7600e+01 <= 1e3*eps       = 1.0000e-02
0 : maxIter   =      20    <= iter          =      3
------------------------- DONE! -------------------------
CPU times: user 30.8 s, sys: 17.5 s, total: 48.2 s
Wall time: 11.1 s
```

</details>


With `54da8d025b00bba9b7641f1349f5cee1397fc891`:

<details>

```
SimPEG.InvProblem will set Regularization.reference_model to m0.
SimPEG.InvProblem will set Regularization.reference_model to m0.
SimPEG.InvProblem will set Regularization.reference_model to m0.
SimPEG.InvProblem will set Regularization.reference_model to m0.
SimPEG.InvProblem will set Regularization.reference_model to m0.
SimPEG.InvProblem will set Regularization.reference_model to m0.
SimPEG.InvProblem will set Regularization.reference_model to m0.

                    SimPEG.InvProblem is setting bfgsH0 to the inverse of the eval2Deriv.
                    ***Done using the default solver Pardiso and no solver_opts.***
                    
model has any nan: 0
=============================== Projected GNCG ===============================
  #     beta     phi_d     phi_m       f      |proj(x-g)-x|  LS    Comment   
-----------------------------------------------------------------------------
x0 has any nan: 0
   0  1.50e-04  4.88e+03  0.00e+00  4.88e+03    1.44e+02      0              
   1  5.01e-05  2.28e+03  5.93e+06  2.58e+03    1.36e+02      0              
   2  1.67e-05  1.41e+03  1.60e+07  1.68e+03    1.18e+02      0   Skip BFGS  
------------------------- STOP! -------------------------
1 : |fc-fOld| = 0.0000e+00 <= tolF*(1+|f0|) = 4.8822e+02
0 : |xc-x_last| = 4.5579e-01 <= tolX*(1+|x0|) = 1.0000e-01
0 : |proj(x-g)-x|    = 1.1765e+02 <= tolG          = 1.0000e-01
0 : |proj(x-g)-x|    = 1.1765e+02 <= 1e3*eps       = 1.0000e-02
0 : maxIter   =      20    <= iter          =      3
------------------------- DONE! -------------------------
CPU times: user 32.5 s, sys: 19.4 s, total: 51.9 s
Wall time: 11.6 s
```

</details>

lheagy commented on SimPEG/regularization/cross_gradient.py:
```diff
@@ -288,12 +286,15 @@ def deriv(self, model):
-            - (((Av @ (g_m1 * g_m2)) @ Av) * g_m1) @ G,
-        ]
+        return (
+            2
+            * np.r_[
+                (((Av @ g_m2**2) @ Av) * g_m1) @ G
+                - (((Av @ (g_m1 * g_m2)) @ Av) * g_m2) @ G,
+                (((Av @ g_m1**2) @ Av) * g_m2) @ G
+                - (((Av @ (g_m1 * g_m2)) @ Av) * g_m1) @ G,
+            ]
```
I will add as a comment

lheagy commented:
# Comments from PR review 

Thanks all! 

@santisoler: I am going through your other comments in [this thread](https://github.com/simpeg/simpeg/pull/1326#pullrequestreview-1822219631) and will respond one at a time  

I believe the first one has been resolved:  
https://github.com/simpeg/simpeg/blob/54da8d025b00bba9b7641f1349f5cee1397fc891/SimPEG/regularization/pgi.py#L840-L849

Nice catch for the examples in [simpeg/tests/em/em1d/test_EM1D_FD_jac_layers.py](https://github.com/simpeg/simpeg/blob/eef5e70557d2f9407485bef0542863c9d13ef43f/tests/em/em1d/test_EM1D_FD_jac_layers.py#L158)! Here, I updated the ones where we are computing the misfit to be consistent with removing the factor of 0.5. The ` dm = m_1D * 0.5 ` lines are choosing a perturbation size, so they are not affecting the derivative. I left those as is. 

And I added comments for the cross gradient. 
 
# Spot check on examples 

I built the docs locally and compared to what we have live. The ones I checked that agree are: 
- https://docs.simpeg.xyz/content/tutorials/04-magnetics/plot_inv_2a_magnetics_induced.html
- https://docs.simpeg.xyz/content/tutorials/06-ip/plot_inv_2_dcip2d.html#sphx-glr-content-tutorials-06-ip-plot-inv-2-dcip2d-py
- https://docs.simpeg.xyz/content/tutorials/13-joint_inversion/plot_inv_3_cross_gradient_pf.html
- https://docs.simpeg.xyz/content/tutorials/14-pgi/plot_inv_1_joint_pf_pgi_full_info_tutorial.html#sphx-glr-content-tutorials-14-pgi-plot-inv-1-joint-pf-pgi-full-info-tutorial-py
- https://docs.simpeg.xyz/content/tutorials/14-pgi/plot_inv_2_joint_pf_pgi_no_info_tutorial.html

This covers L2, sparse norms, cross gradient and PGI. So I think we are in good shape!

lheagy commented on SimPEG/regularization/cross_gradient.py:
```diff
@@ -404,4 +405,4 @@ def deriv2(self, model, v=None):
                     - (Av.T @ (Av @ (g_m2 * g_m1))) * Gv1
                 )
-            return np.r_[p1, p2]
+            return 2 * np.r_[p1, p2]
```
added a comment

santisoler commented:
Thanks @lheagy for the changes and the reply.

I just edited my review comment, striking out the suggestions that were already solved:
- all the changes in the PGI were already covered by @thibaut-kobold 
- you updated the bits in [simpeg/examples/20-published/plot_tomo_joint_with_volume.py](https://github.com/simpeg/simpeg/blob/eef5e70557d2f9407485bef0542863c9d13ef43f/examples/20-published/plot_tomo_joint_with_volume.py#L39-L83)
- and updated the tests for EM 1D.

I caught a minor mistake in  the docstrings of some regularizations that I'll fix in a separate PR.

If everything passes, I think this PR should be ready to go. Is there anything else you would like to take a second loot at?

lheagy commented:
Thanks @santisoler! This is good to go from my perspective
